{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/bin/python\r\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/bin/sh: line 0: source: activate: file not found\r\n"
     ]
    }
   ],
   "source": [
    "!source activate py36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "#sys.path.append('/home/faske/g/anaconda3/envs/py34/lib/python3.4/site-packages')\n",
    "sys.path.append('/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages')\n",
    "sys.path.append(\"/data/gpfs/assoc/parchmanlab/tfaske/ipynb/include_utils\")\n",
    "\n",
    "import ipyparallel as ipp\n",
    "import os, time\n",
    "import include_utils as u\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import random\n",
    "import numbers\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "#import vcf\n",
    "from sklearn import preprocessing\n",
    "from subprocess import Popen, PIPE, call, check_output\n",
    "import seaborn as sns\n",
    "from IPython.display import FileLink\n",
    "import urllib.request as urllib2\n",
    "import dill\n",
    "import traceback\n",
    "from pandas import Series, DataFrame\n",
    "import gzip\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore',category=pd.io.pytables.PerformanceWarning)\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "from Bio import SeqIO\n",
    "#import pysam\n",
    "from collections import OrderedDict, namedtuple, Counter\n",
    "import operator\n",
    "import multiprocessing as mp\n",
    "import shutil\n",
    "import tempfile\n",
    "#from ipyparallel import Client\n",
    "import scandir\n",
    "import glob\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "import pickle\n",
    "import re\n",
    "from itertools import chain\n",
    "#import Levenshtein as lv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfutils = \"vcfutils.pl\"\n",
    "vcftools = \"vcftools\"\n",
    "bcftools = \"bcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\"\n",
    "java = \"java\"\n",
    "plink = \"plink --noweb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = '/data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/tfaske/sagebrush/denovo/SNPcall/test/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/fb_cat.vcf.gz'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"fb_cat.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter first round \n",
    "    missing data, MAF, biallelic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/fb_cat.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--thin 100\n",
      "\t--max-missing 0.6\n",
      "\t--out fb_miss60_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 55771 out of a possible 503227 Sites\n",
      "Run Time = 9.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--max-missing 0.6 \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'fb_miss60_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"%s.recode.vcf\" % (analysis_dir + 'fb_miss60_thin100_MAF2')\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove bad_indv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/fb_miss60_thin100_MAF2.recode.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/fb_miss60_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz --out $vcf_filtered_gz --missing-indv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_imiss(vcf_file):\n",
    "    imiss_file = !ls {vcf_file}.imiss\n",
    "    imiss_df = pd.read_csv(imiss_file[0], sep=\"\\t\")\n",
    "    imiss_df.index = imiss_df.INDV\n",
    "    return imiss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INDV</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AT2_EW_36</th>\n",
       "      <td>AT2_EW_36</td>\n",
       "      <td>55771</td>\n",
       "      <td>0</td>\n",
       "      <td>5348</td>\n",
       "      <td>0.095892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT2_HC_2</th>\n",
       "      <td>AT2_HC_2</td>\n",
       "      <td>55771</td>\n",
       "      <td>0</td>\n",
       "      <td>6655</td>\n",
       "      <td>0.119327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT2_VM_5</th>\n",
       "      <td>AT2_VM_5</td>\n",
       "      <td>55771</td>\n",
       "      <td>0</td>\n",
       "      <td>10004</td>\n",
       "      <td>0.179376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT2_YB_79</th>\n",
       "      <td>AT2_YB_79</td>\n",
       "      <td>55771</td>\n",
       "      <td>0</td>\n",
       "      <td>5674</td>\n",
       "      <td>0.101737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT4_DS_2</th>\n",
       "      <td>AT4_DS_2</td>\n",
       "      <td>55771</td>\n",
       "      <td>0</td>\n",
       "      <td>6639</td>\n",
       "      <td>0.119040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                INDV  N_DATA  N_GENOTYPES_FILTERED  N_MISS    F_MISS\n",
       "INDV                                                                \n",
       "AT2_EW_36  AT2_EW_36   55771                     0    5348  0.095892\n",
       "AT2_HC_2    AT2_HC_2   55771                     0    6655  0.119327\n",
       "AT2_VM_5    AT2_VM_5   55771                     0   10004  0.179376\n",
       "AT2_YB_79  AT2_YB_79   55771                     0    5674  0.101737\n",
       "AT4_DS_2    AT4_DS_2   55771                     0    6639  0.119040"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df = get_imiss(vcf_filtered_gz)\n",
    "imiss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    15.000000\n",
       "mean      0.111447\n",
       "std       0.041134\n",
       "min       0.042818\n",
       "25%       0.086586\n",
       "50%       0.114432\n",
       "75%       0.130722\n",
       "max       0.190637\n",
       "Name: F_MISS, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df.F_MISS.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15, 0, 0, 0)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(imiss_df),len(imiss_df[imiss_df.F_MISS >= .4]),len(imiss_df[imiss_df.F_MISS >= .5]), len(imiss_df[imiss_df.F_MISS >= .25])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_indv = imiss_df.INDV[imiss_df.F_MISS >= .4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"bad_indv.txt\"), \"w\") as o:\n",
    "    o.write(\"INDV\\n\")\n",
    "    for elem in bad_indv.index:\n",
    "        o.write(\"%s\\n\" % elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/fb_miss60_thin100_MAF2.recode.vcf.gz\n",
      "\t--remove /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/bad_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 5.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"bad_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"snps_indv_removed\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_bad_remove = os.path.join(analysis_dir, \"snps_indv_removed.recode.vcf\")\n",
    "vcf_bad_remove_gz = vcf_bad_remove + \".gz\"\n",
    "!$bgzip -c {vcf_bad_remove} > {vcf_bad_remove_gz}\n",
    "!$tabix {vcf_bad_remove_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Quality for Each Site\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 0.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n",
      "\tHWE: Only using fully diploid SNPs.\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "\tIndividual Heterozygosity: Only using fully diploid SNPs.\n",
      "After filtering, kept 55771 out of a possible 55771 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_bad_remove_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bad_remove_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:65: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'snps_indv_removed.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>...</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_10000-23</th>\n",
       "      <td>979</td>\n",
       "      <td>82465</td>\n",
       "      <td>65.26670</td>\n",
       "      <td>1326.35000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.73230</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017836</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.966667</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_100004-25</th>\n",
       "      <td>1204</td>\n",
       "      <td>104028</td>\n",
       "      <td>80.26670</td>\n",
       "      <td>527.63800</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5829.85000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>-0.933333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_100010-6</th>\n",
       "      <td>75</td>\n",
       "      <td>851</td>\n",
       "      <td>6.25000</td>\n",
       "      <td>34.75000</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>76.26020</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.958333</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_100024-5</th>\n",
       "      <td>467</td>\n",
       "      <td>23241</td>\n",
       "      <td>31.13330</td>\n",
       "      <td>621.55200</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>136.38400</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.674330</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>-0.160000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_100034-5</th>\n",
       "      <td>77</td>\n",
       "      <td>561</td>\n",
       "      <td>5.13333</td>\n",
       "      <td>11.83810</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>61.65040</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076531</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.965517</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>-0.035714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_99965-22</th>\n",
       "      <td>164</td>\n",
       "      <td>2578</td>\n",
       "      <td>12.61540</td>\n",
       "      <td>42.42310</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>100.04900</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.961538</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_99971-4</th>\n",
       "      <td>109</td>\n",
       "      <td>1011</td>\n",
       "      <td>7.78571</td>\n",
       "      <td>12.48900</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>698.96300</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.642857</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_99987-7</th>\n",
       "      <td>74</td>\n",
       "      <td>732</td>\n",
       "      <td>7.40000</td>\n",
       "      <td>20.48890</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>15.24600</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_99992-28</th>\n",
       "      <td>240</td>\n",
       "      <td>5430</td>\n",
       "      <td>17.14290</td>\n",
       "      <td>101.20900</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>91.64640</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_99993-73</th>\n",
       "      <td>51</td>\n",
       "      <td>319</td>\n",
       "      <td>4.63636</td>\n",
       "      <td>8.25455</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>3.32974</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.954545</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>-9.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55771 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH   VAR_DEPTH  \\\n",
       "dDocent_Contig_10000-23         979        82465    65.26670  1326.35000   \n",
       "dDocent_Contig_100004-25       1204       104028    80.26670   527.63800   \n",
       "dDocent_Contig_100010-6          75          851     6.25000    34.75000   \n",
       "dDocent_Contig_100024-5         467        23241    31.13330   621.55200   \n",
       "dDocent_Contig_100034-5          77          561     5.13333    11.83810   \n",
       "...                             ...          ...         ...         ...   \n",
       "dDocent_Contig_99965-22         164         2578    12.61540    42.42310   \n",
       "dDocent_Contig_99971-4          109         1011     7.78571    12.48900   \n",
       "dDocent_Contig_99987-7           74          732     7.40000    20.48890   \n",
       "dDocent_Contig_99992-28         240         5430    17.14290   101.20900   \n",
       "dDocent_Contig_99993-73          51          319     4.63636     8.25455   \n",
       "\n",
       "                          N_DATA  N_GENOTYPE_FILTERED  N_MISS    F_MISS  \\\n",
       "dDocent_Contig_10000-23       30                    0       0  0.000000   \n",
       "dDocent_Contig_100004-25      30                    0       0  0.000000   \n",
       "dDocent_Contig_100010-6       27                    0       3  0.111111   \n",
       "dDocent_Contig_100024-5       30                    0       0  0.000000   \n",
       "dDocent_Contig_100034-5       30                    0       0  0.000000   \n",
       "...                          ...                  ...     ...       ...   \n",
       "dDocent_Contig_99965-22       28                    0       2  0.071429   \n",
       "dDocent_Contig_99971-4        29                    0       1  0.034483   \n",
       "dDocent_Contig_99987-7        25                    0       5  0.200000   \n",
       "dDocent_Contig_99992-28       29                    0       1  0.034483   \n",
       "dDocent_Contig_99993-73       26                    0       4  0.153846   \n",
       "\n",
       "                                QUAL N_ALLELES  ...  ChiSq_HWE     P_HWE  \\\n",
       "dDocent_Contig_10000-23     11.73230         2  ...   0.017836  1.000000   \n",
       "dDocent_Contig_100004-25  5829.85000         2  ...  15.000000  0.000211   \n",
       "dDocent_Contig_100010-6     76.26020         2  ...        NaN       NaN   \n",
       "dDocent_Contig_100024-5    136.38400         2  ...   0.600000  1.000000   \n",
       "dDocent_Contig_100034-5     61.65040         2  ...   0.076531  1.000000   \n",
       "...                              ...       ...  ...        ...       ...   \n",
       "dDocent_Contig_99965-22    100.04900         2  ...        NaN       NaN   \n",
       "dDocent_Contig_99971-4     698.96300         2  ...        NaN       NaN   \n",
       "dDocent_Contig_99987-7      15.24600         2  ...        NaN       NaN   \n",
       "dDocent_Contig_99992-28     91.64640         2  ...        NaN       NaN   \n",
       "dDocent_Contig_99993-73      3.32974         2  ...        NaN       NaN   \n",
       "\n",
       "                         P_HET_DEFICIT P_HET_EXCESS A1_allele  A2_allele  \\\n",
       "dDocent_Contig_10000-23            1.0     1.000000         T          G   \n",
       "dDocent_Contig_100004-25           1.0     0.000211         C          T   \n",
       "dDocent_Contig_100010-6            NaN          NaN         C          T   \n",
       "dDocent_Contig_100024-5            1.0     0.674330         G          A   \n",
       "dDocent_Contig_100034-5            1.0     0.965517         C          T   \n",
       "...                                ...          ...       ...        ...   \n",
       "dDocent_Contig_99965-22            NaN          NaN         C          A   \n",
       "dDocent_Contig_99971-4             NaN          NaN         G          T   \n",
       "dDocent_Contig_99987-7             NaN          NaN         G          T   \n",
       "dDocent_Contig_99992-28            NaN          NaN         C          A   \n",
       "dDocent_Contig_99993-73            NaN          NaN         T          C   \n",
       "\n",
       "                           A1_freq   A2_freq       MAF       Fis  \n",
       "dDocent_Contig_10000-23   0.966667  0.033333  0.033333  0.000000  \n",
       "dDocent_Contig_100004-25  0.500000  0.500000  0.500000 -0.933333  \n",
       "dDocent_Contig_100010-6   0.958333  0.041667  0.041667 -9.000000  \n",
       "dDocent_Contig_100024-5   0.833333  0.166667  0.166667 -0.160000  \n",
       "dDocent_Contig_100034-5   0.933333  0.066667  0.066667 -0.035714  \n",
       "...                            ...       ...       ...       ...  \n",
       "dDocent_Contig_99965-22   0.961538  0.038462  0.038462 -9.000000  \n",
       "dDocent_Contig_99971-4    0.642857  0.357143  0.357143 -9.000000  \n",
       "dDocent_Contig_99987-7    0.950000  0.050000  0.050000 -9.000000  \n",
       "dDocent_Contig_99992-28   0.928571  0.071429  0.071429 -9.000000  \n",
       "dDocent_Contig_99993-73   0.954545  0.045455  0.045455 -9.000000  \n",
       "\n",
       "[55771 rows x 25 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_469-22</th>\n",
       "      <td>dDocent_Contig_469</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>T:0.9</td>\n",
       "      <td>C:0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_2655-87</th>\n",
       "      <td>dDocent_Contig_2655</td>\n",
       "      <td>87</td>\n",
       "      <td>2</td>\n",
       "      <td>24</td>\n",
       "      <td>A:0.916667</td>\n",
       "      <td>C:0.0833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_2973-21</th>\n",
       "      <td>dDocent_Contig_2973</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>C:0.5</td>\n",
       "      <td>T:0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_3819-42</th>\n",
       "      <td>dDocent_Contig_3819</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>T:0.5</td>\n",
       "      <td>A:0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_3966-63</th>\n",
       "      <td>dDocent_Contig_3966</td>\n",
       "      <td>63</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>T:0.866667</td>\n",
       "      <td>A:0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_448372-7</th>\n",
       "      <td>dDocent_Contig_448372</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>A:0.964286</td>\n",
       "      <td>G:0.0357143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_450674-2</th>\n",
       "      <td>dDocent_Contig_450674</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>C:0.714286</td>\n",
       "      <td>T:0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_458652-4</th>\n",
       "      <td>dDocent_Contig_458652</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>24</td>\n",
       "      <td>G:0.958333</td>\n",
       "      <td>A:0.0416667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_460788-32</th>\n",
       "      <td>dDocent_Contig_460788</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>A:0.954545</td>\n",
       "      <td>G:0.0454545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_482580-21</th>\n",
       "      <td>dDocent_Contig_482580</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>T:0.95</td>\n",
       "      <td>C:0.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55771 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM POS N_ALLELES N_CHR  \\\n",
       "dDocent_Contig_469-22        dDocent_Contig_469  22         2    30   \n",
       "dDocent_Contig_2655-87      dDocent_Contig_2655  87         2    24   \n",
       "dDocent_Contig_2973-21      dDocent_Contig_2973  21         2    30   \n",
       "dDocent_Contig_3819-42      dDocent_Contig_3819  42         2    30   \n",
       "dDocent_Contig_3966-63      dDocent_Contig_3966  63         2    30   \n",
       "...                                         ...  ..       ...   ...   \n",
       "dDocent_Contig_448372-7   dDocent_Contig_448372   7         2    28   \n",
       "dDocent_Contig_450674-2   dDocent_Contig_450674   2         2    28   \n",
       "dDocent_Contig_458652-4   dDocent_Contig_458652   4         2    24   \n",
       "dDocent_Contig_460788-32  dDocent_Contig_460788  32         2    22   \n",
       "dDocent_Contig_482580-21  dDocent_Contig_482580  21         2    20   \n",
       "\n",
       "                             A1_FREQ      A2_FREQ  \n",
       "dDocent_Contig_469-22          T:0.9        C:0.1  \n",
       "dDocent_Contig_2655-87    A:0.916667  C:0.0833333  \n",
       "dDocent_Contig_2973-21         C:0.5        T:0.5  \n",
       "dDocent_Contig_3819-42         T:0.5        A:0.5  \n",
       "dDocent_Contig_3966-63    T:0.866667   A:0.133333  \n",
       "...                              ...          ...  \n",
       "dDocent_Contig_448372-7   A:0.964286  G:0.0357143  \n",
       "dDocent_Contig_450674-2   C:0.714286   T:0.285714  \n",
       "dDocent_Contig_458652-4   G:0.958333  A:0.0416667  \n",
       "dDocent_Contig_460788-32  A:0.954545  G:0.0454545  \n",
       "dDocent_Contig_482580-21      T:0.95       C:0.05  \n",
       "\n",
       "[55771 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_469-22</th>\n",
       "      <td>dDocent_Contig_469</td>\n",
       "      <td>22</td>\n",
       "      <td>12/3/0</td>\n",
       "      <td>12.15/2.70/0.15</td>\n",
       "      <td>0.185185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.896552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_2973-21</th>\n",
       "      <td>dDocent_Contig_2973</td>\n",
       "      <td>21</td>\n",
       "      <td>0/15/0</td>\n",
       "      <td>3.75/7.50/3.75</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.000211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_3819-42</th>\n",
       "      <td>dDocent_Contig_3819</td>\n",
       "      <td>42</td>\n",
       "      <td>0/15/0</td>\n",
       "      <td>3.75/7.50/3.75</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.000211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_3966-63</th>\n",
       "      <td>dDocent_Contig_3966</td>\n",
       "      <td>63</td>\n",
       "      <td>11/4/0</td>\n",
       "      <td>11.27/3.47/0.27</td>\n",
       "      <td>0.355030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.796935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_4271-10</th>\n",
       "      <td>dDocent_Contig_4271</td>\n",
       "      <td>10</td>\n",
       "      <td>0/15/0</td>\n",
       "      <td>3.75/7.50/3.75</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.000211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_336302-5</th>\n",
       "      <td>dDocent_Contig_336302</td>\n",
       "      <td>5</td>\n",
       "      <td>0/14/1</td>\n",
       "      <td>3.27/7.47/4.27</td>\n",
       "      <td>11.484380</td>\n",
       "      <td>0.001734</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.001690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_351207-21</th>\n",
       "      <td>dDocent_Contig_351207</td>\n",
       "      <td>21</td>\n",
       "      <td>11/3/1</td>\n",
       "      <td>10.42/4.17/0.42</td>\n",
       "      <td>1.176000</td>\n",
       "      <td>0.325670</td>\n",
       "      <td>0.32567</td>\n",
       "      <td>0.980843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_383096-28</th>\n",
       "      <td>dDocent_Contig_383096</td>\n",
       "      <td>28</td>\n",
       "      <td>14/1/0</td>\n",
       "      <td>14.02/0.97/0.02</td>\n",
       "      <td>0.017836</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_409231-6</th>\n",
       "      <td>dDocent_Contig_409231</td>\n",
       "      <td>6</td>\n",
       "      <td>12/3/0</td>\n",
       "      <td>12.15/2.70/0.15</td>\n",
       "      <td>0.185185</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.896552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_416321-4</th>\n",
       "      <td>dDocent_Contig_416321</td>\n",
       "      <td>4</td>\n",
       "      <td>14/1/0</td>\n",
       "      <td>14.02/0.97/0.02</td>\n",
       "      <td>0.017836</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24456 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM  POS OBS(HOM1/HET/HOM2)  \\\n",
       "dDocent_Contig_469-22        dDocent_Contig_469   22             12/3/0   \n",
       "dDocent_Contig_2973-21      dDocent_Contig_2973   21             0/15/0   \n",
       "dDocent_Contig_3819-42      dDocent_Contig_3819   42             0/15/0   \n",
       "dDocent_Contig_3966-63      dDocent_Contig_3966   63             11/4/0   \n",
       "dDocent_Contig_4271-10      dDocent_Contig_4271   10             0/15/0   \n",
       "...                                         ...  ...                ...   \n",
       "dDocent_Contig_336302-5   dDocent_Contig_336302    5             0/14/1   \n",
       "dDocent_Contig_351207-21  dDocent_Contig_351207   21             11/3/1   \n",
       "dDocent_Contig_383096-28  dDocent_Contig_383096   28             14/1/0   \n",
       "dDocent_Contig_409231-6   dDocent_Contig_409231    6             12/3/0   \n",
       "dDocent_Contig_416321-4   dDocent_Contig_416321    4             14/1/0   \n",
       "\n",
       "                         E(HOM1/HET/HOM2)  ChiSq_HWE     P_HWE  P_HET_DEFICIT  \\\n",
       "dDocent_Contig_469-22     12.15/2.70/0.15   0.185185  1.000000        1.00000   \n",
       "dDocent_Contig_2973-21     3.75/7.50/3.75  15.000000  0.000211        1.00000   \n",
       "dDocent_Contig_3819-42     3.75/7.50/3.75  15.000000  0.000211        1.00000   \n",
       "dDocent_Contig_3966-63    11.27/3.47/0.27   0.355030  1.000000        1.00000   \n",
       "dDocent_Contig_4271-10     3.75/7.50/3.75  15.000000  0.000211        1.00000   \n",
       "...                                   ...        ...       ...            ...   \n",
       "dDocent_Contig_336302-5    3.27/7.47/4.27  11.484380  0.001734        1.00000   \n",
       "dDocent_Contig_351207-21  10.42/4.17/0.42   1.176000  0.325670        0.32567   \n",
       "dDocent_Contig_383096-28  14.02/0.97/0.02   0.017836  1.000000        1.00000   \n",
       "dDocent_Contig_409231-6   12.15/2.70/0.15   0.185185  1.000000        1.00000   \n",
       "dDocent_Contig_416321-4   14.02/0.97/0.02   0.017836  1.000000        1.00000   \n",
       "\n",
       "                          P_HET_EXCESS  \n",
       "dDocent_Contig_469-22         0.896552  \n",
       "dDocent_Contig_2973-21        0.000211  \n",
       "dDocent_Contig_3819-42        0.000211  \n",
       "dDocent_Contig_3966-63        0.796935  \n",
       "dDocent_Contig_4271-10        0.000211  \n",
       "...                                ...  \n",
       "dDocent_Contig_336302-5       0.001690  \n",
       "dDocent_Contig_351207-21      0.980843  \n",
       "dDocent_Contig_383096-28      1.000000  \n",
       "dDocent_Contig_409231-6       0.896552  \n",
       "dDocent_Contig_416321-4       1.000000  \n",
       "\n",
       "[24456 rows x 8 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    5.577100e+04\n",
       "mean     5.103283e+02\n",
       "std      1.088253e+04\n",
       "min      9.000000e+00\n",
       "25%      5.300000e+01\n",
       "50%      9.300000e+01\n",
       "75%      2.150000e+02\n",
       "max      2.485918e+06\n",
       "Name: SUM_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    5.577100e+04\n",
       "mean     1.950249e+03\n",
       "std      5.863799e+04\n",
       "min      7.959420e-08\n",
       "25%      3.006505e+01\n",
       "50%      8.368840e+01\n",
       "75%      2.819060e+02\n",
       "max      1.289440e+07\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAIJCAYAAAC2k+IhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nOzde5wlZX3v+89XCCAXuUUjijmjKMJRjIImOkZEOTEkIWJ0PGK8oHjZGsAtgieJEgWDiXuLoICRbHZgVJKg4kb3KKImgIDjDdCNRgQFRqPiBYFBYMAAv/NHVevKonume6ZmnllrPu/Xa72KVfXcVk3T3d+ueupJVSFJkiRJG9r9Wg9AkiRJ0qbJMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmti89QC0/iS5HngAsKLxUCRJkjS9FgG3VtXDF1rRMDLdHnD/+99/pz333HOn1gORJEnSdLrqqqtYtWrVWtU1jEy3FXvuuedOl19+eetxSJIkaUrts88+XHHFFSvWpq5zRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUxOatB6DptOgvPtl6CBvcinf8UeshSJIkTRSvjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpqYuDCSZEWSmuP1oznqLE5yXpKbkqxKcmWS1yfZbDX9HJjkoiQrk9yW5EtJDlnD2A5J8uW+/Mq+/oGrKb9ZkiP78azqx3deksXzPyOSJEnSZNq89QDW0krg3bPsv218R5KDgI8CdwIfAm4C/hg4CXgq8PxZ6hwOnAL8DDgL+AWwBFiaZK+qOnqWOicARwHfB04HtgAOBpYlOaKqTh0rH+Dsvt2rgVOBnYAXABcneV5VfXyNZ0KSJEmaUJMaRm6pqmPXVCjJA+iCwT3AflV1Wb//r4ALgCVJDq6qs0fqLAJOoAstT6yqFf3+twFfAY5K8tGq+sJIncV0QeRa4ElVdXO//53A5cAJST4x01bvYLogshzYv6ru7OucBlwKnJ7kgqr6+cJOjSRJkjQZJu42rQVaAjwQOHsmiAD0v/gf07997VidQ4EtgVNHw0MfMP6mf/uasToz798+E0T6OiuA9/btvXyszky/x8wEkb7OV+iu4DywH78kSZI0lSY1jGyZ5MVJ3pTkvyZ5xhzzP57Zb8+f5djFwB3A4iRbzrPOp8bKrFWdJFsBi/v+L1lAP5IkSdLUmNTbtB4MfHBs3/VJXl5VnxvZ9+h+e814A1V1d5LrgccAjwCumkedG5LcDuyaZOuquiPJNsBDgduq6oZZxvrtfrv7yL7dgM2A66rq7nnWmVOSy+c4tMd86kuSJEktTOKVkTOB/ekCyTbAXsDfA4uATyX5rZGy2/fblXO0NbN/h7Wos/3Ydn30scMcxyVJkqSJN3FXRqrquLFd3wBek+Q2uknkxwJ/sqHH1VJV7TPb/v6Kyd4beDiSJEnSvEzilZG5nNZv9x3ZN34VY9zM/lvWos7Kse366OOWOY5LkiRJE2+awshP++02I/uu7rf3mXuRZHPg4cDdwHXzrLNL3/73q+oOgKq6HfgBsG1/fNyj+u3oHJRr6R43/Ih+HPOpI0mSJE2VaQojT+63o8Hign57wCzl9wW2BpZX1V3zrPMHY2XWqk7/KN/lff9PW0A/kiRJ0tSYqDCSZM/+6VXj+xfRrWAO3YrpM84BbgQOTvLEkfJbAcf3b9831tyZwF3A4X27M3V2BN7Uvz1trM7M+zf35UbHdVjf3pljdWb6Pb4fz0ydJ9Gtwv5TupXjJUmSpKk0aRPYX0C3AvrFwHeBn9M9JvePgK2A8+hWTwegqm5N8iq6UHJRkrPpVlZ/Nt0jfM+hW2CQkTrXJ3kjcDJwWZIPAb+gW4BwV+Bdo6uv93WWJzkReANwZZJzgC368e4EHDG2+jrA2cBz+3a/mmQZsHNfZzPgVVV169qeKEmSJGljN2lh5EK6EPEE4Kl08zduAS6lW3fkg1VVoxWq6mNJng68GXgeXWj5Dl1wOHm8fF/nlCQrgKOBl9JdQfom3Wrp759tYFV1VJKv010JeTVwL3AF8M6q+sQs5SvJC+lu1zoUOAK4k24xxuOravkCzoskSZI0cSYqjPQLGn5ujQXvW+/zwB8usM4yYNkC6ywFli6g/N3ASf1LkiRJ2qRM1JwRSZIkSdPDMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmpj4MJLkxUmqf71yjjIHJrkoycoktyX5UpJD1tDuIUm+3Jdf2dc/cDXlN0tyZJIrk6xKclOS85IsXk2d+yc5LsnVSe5M8pMkH06y5/zPgCRJkjSZJjqMJHkYcCpw22rKHA4sAx4LnAWcDjwEWJrkhDnqnAAsBXbpy58F7AUs69sbLx/gbOBEYIt+TOcC+wIXJzloljpbAp8F3gLcCrwH+BfgT4DLkvzOGk+AJEmSNME2bz2AtdUHgDOBnwH/Czh6ljKLgBOAm4AnVtWKfv/bgK8ARyX5aFV9YaTOYuAo4FrgSVV1c7//ncDlwAlJPjHTVu9gYAmwHNi/qu7s65wGXAqcnuSCqvr5SJ03AE8FzgFeUFX39nU+BHwMOCPJXjP7JUmSpGkzyVdGXgc8E3g5cPscZQ4FtgROHQ0PfcD4m/7ta8bqzLx/+0wQ6eusAN7bt/fysTqv7bfHzASRvs5XgA8BD6QLK8Avg9RMP//faOCoqo8DlwD/N/D0OT6XJEmSNPEmMoz0cyreAbynqi5eTdFn9tvzZzn2qbEya1UnyVbAYuAOuhAxn352A34TuKaqrl/A2CRJkqSpMXG3aSXZHPgg8D3gTWso/uh+e834gaq6IcntwK5Jtq6qO5JsAzwUuK2qbpilvW/3291H9u0GbAZcV1V3z7POnONaTZ05Jbl8jkN7zKe+JEmS1MLEhRG6Cd9PAH63qlatoez2/XblHMdXAtv05e6YZ3mAHRbYxxB1JEmSpKkyUWGkf8LUm4B3jU4639RV1T6z7e+vmOy9gYcjSZIkzcvEzBnpb8/6AN2tTX81z2ozVxi2n+P4+BWK+Za/ZS36WNc6kiRJ0lSZmDACbEs3h2JP4M6RhQ4LeGtf5vR+37v791f32/vMvUiyC90tWt+vqjsAqup24AfAtv3xcY/qt6NzPa4F7gEe0Qem+dSZc1yrqSNJkiRNlUm6Tesu4B/mOLY33TySS+l+0Z+5hesCurU8DhjZN+MPRsqMugB4SV/nzDXVqao7kywHnta/LpxHP9fSTcDfPcnDZ3mi1lxjkyRJkqbGxFwZqapVVfXK2V7A/+6Lvb/f96H+/Zl0IebwfgFEAJLsyK+exHXaWFcz79/cl5upswg4rG9vPKS8r98e3z/qd6bOk4AXAD8FPjryWWqkn/+e5H4jdQ6iCzXfBD439xmRJEmSJtskXRlZsKq6PskbgZOBy/rVzX9BtwDhrswyEb6qlic5kW6F9CuTnANsQRcqdgKOGFt9HeBs4Ll9u19NsgzYua+zGfCqqrp1rM6JwIF9nS8l+Ve6tUeeT/dkr0NdfV2SJEnTbKrDCEBVnZJkBXA08FK6q0HfpFst/f1z1DkqydfproS8GrgXuAJ4Z1V9YpbyleSFwHK6Vd+PAO4ELgaOr6rls9S5K8nvAX8BvBA4ErgV+Bjw1qr65jp9cEmSJGkjNxVhpKqOBY5dzfFlwLIFtrkUWLqA8ncDJ/Wv+da5g27dlLcsZGySJEnSNJiYOSOSJEmSpothRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4OGkSS/NmR7kiRJkqbX0FdGfpDkvyV55MDtSpIkSZoyQ4eR+wFvBK5O8tkkz0uy2cB9SJIkSZoCQ4eRhwAvBi4B9gc+DHw/yduTLBq4L0mSJEkTbNAwUlW/qKp/qqr9gD2AdwObA38JfCfJeUkOSuLEeUmSJGkTt95CQVVdU1VHAQ/lV1dLDgD+F/C9JMcmecj66l+SJEnSxm29X6Goql8AnwTOBX4IhO52rrcA1yd5d5It1/c4JEmSJG1c1msYSfLkJGfShZCTgG2Ak4HHA4cCVwNH0N3OJUmSJGkTsvnQDSbZDngJ8F+Ax9JdCfkq8HfAP1XVqr7olUk+CJwPLAFeO/RYJEmSJG28Bg0jSf4B+H+BrYG7gA8Cf1dVX56tfFXdk+Qi4JlDjkOSJEnSxm/oKyMvB64FTgPOrKqb5lHnIuBtA49DkiRJ0kZu6DByQFV9ZiEVqurzwOcHHockSZKkjdzQ64wsKIhIkiRJ2nQNGkaS7J/kjLnWD0nykP74fkP2K0mSJGnyDH2b1hHAHlX1w9kOVtUPkzwF2J5urogkSZKkTdTQ64zsDSxfQ5lLgScO3K8kSZKkCTN0GHkQ3QKHq/PjvpwkSZKkTdjQYWQl8LA1lHkYcPvA/UqSJEmaMEOHkS8Dz0ny4NkO9hPbn9OXkyRJkrQJGzqMnAJsB1yS5NlJtgRIsmWSg4CLgW2BkwfuV5IkSdKEGfRpWlX1mSR/DfwVcC5QSW4GdgTSv/66qs4fsl9JkiRJk2foKyNU1VuBA4DzgJvoHuN7E/BJ4Pf745IkSZI2cUOvMwL8ciV2V2OXJEmSNKfBr4xIkiRJ0nyslysjAEm2ppsrstlsx6vqe+urb0mSJEkbv8HDSJKXAH8O7LmaYrU++pYkSZI0OQYNBEleBpwB3ANcAvw7cPeQfUiSJEmaDkNfnTgauBn43aq6auC2JUmSJE2RoSewPxL4iEFEkiRJ0poMHUZuAu4auE1JkiRJU2joMPIJYL8kGbhdSZIkSVNm6DDyl8CWwGlJth24bUmSJElTZOgJ7B8B7gBeCfxpkm8Dt8xSrqpq/4H7liRJkjRBhg4j+4389zbA4+coVwP3K0mSJGnCDBpGqmro274kSZIkTSnDgyRJkqQmDCOSJEmSmhg8jCS5X5Ijknwxycokd48ce0KSv0uy+zq0/9+S/GuSf0+yKslNSb6a5K1Jdp6jzuIk5/VlVyW5Msnrk2y2mn4OTHJR/xluS/KlJIesYWyHJPlyX35lX//A1ZTfLMmR/XhmPst5SRbP/4xIkiRJk2nQMJJkC+CzwLuB3YCfA6NrjlwPHAq8aB26OZJucvxngfcA/wjcDRwLXJnkYWNjOgi4GNgXOBc4FdgCOAk4e47PcTiwDHgscBZwOvAQYGmSE+aocwKwFNilL38WsBewrG9vvHz6/k/sx3NqP759gYv7cUuSJElTa+grI28EngEcB/wG8D9HD1bVLXTB4PfXoY8HVNWTq+rQqvqLqjqiqp4E/A1dYPjLmYJJHkAXDO4B9quqV1TVG+me8vUFYEmSg0cbT7IIOIFuNfknVtVhVXUk8DjgWuCoJE8Zq7MYOKo//riqOrKqDgP26ds5oW931MHAEmA58PiqemNVvYLu/N0DnJ5ku3U4T5IkSdJGbegw8iLg81X1tqq6l9kf4Xs98Jtr20FV3TnHoQ/320eN7FsCPBA4u6ouG2vjmP7ta8faOZRu4cZTq2rFSJ2b6QIPwGvG6sy8f3tfbqbOCuC9fXsvH6sz0+8xo5+pqr4CfKgf95JZPqckSZI0FYYOIw8HvriGMjcBOw3cL8Af99srR/Y9s9+eP0v5i+kWaFycZMt51vnUWJm1qpNkK2Bx3/8lC+hHkiRJmhpDL3p4J7DDGsr8JrOvyr4gSY4GtgW2B54I/C5dEHnHSLFH99trxutX1d1JrgceAzwCuGoedW5Icjuwa5Ktq+qOJNsADwVuq6obZhnqt/vt6KT93YDNgOuq6u77Vpm1zpySXD7HoT3mU1+SJElqYegw8jXgWUm2qKpfjB9Msj3dfJHlA/R1NN28lBnnAy+rqp+O7Nu+366co42Z/aMBaj51tunL3bEe+xivI0mSJE2VoW/T+h/Aw4B/7CeP/1KSHeieNrUjcNq6dlRVD66qAA8Gnkt3deOrSfZe17YnTVXtM9sL+FbrsUmSJElzGfTKSFX9c5LfA14GPBu4GSDJZXS3Q20JvLeqzhuwzx8D5ya5gu7Wqg/QPZIXfnWFYfvZ6o7sH71tbCXw6/2xn62mzsqx7UL7WGgdSZIkaaoMvuhhVR1K90Sqb9I9ESrA3sB3gFdU1RFD99n3+92+z8ck+fV+99X99j5zL5JsTjfh/m7gupFDq6uzC90tWt+vqjv6fm8HfgBs2x8fN/N0r9E5KNfSPb73Ef045lNHkiRJmiqDhxGAqlpaVU+gm2C+K7BdVe1VVWeuj/5GPKTf3tNvL+i3B8xSdl9ga2B5Vd01sn91df5grMxa1ekf5bu87/9pC+hHkiRJmhrrJYzMqKpVVfXD/urBOkuyez8Jfnz//ZK8HXgQXbiYWevjHOBG4OAkTxwpvxVwfP/2fWPNnQncBRw+ulBhkh2BN/Vvx+e8zLx/c19ups4i4LC+vfEgNtPv8f14Zuo8CXgB8FPgo+OfVZIkSZoWQz9Na337Q+Bvk1xKt3jiz+ieqPV0ugnsPwJeNVO4qm5N8iq6UHJRkrPp1jl5Nt0jfM+hW2CQkTrXJ3kjcDJwWZIPAb+gW4BwV+BdVfWFsTrLk5wIvAG4Msk5wBZ0oWIn4IjRBRR7Z9NNvF9CN/F+GbBzX2cz4FVVdevanihJkiRpYzdoGEly3ZpLAVBVtdtadPEvwCPp1hR5At2jb2+nm1vxQeDkqrpprKOPJXk68GbgecBWdPNX3tCXv88q8VV1SpIVdI8PfindFaRv0q2W/v45PtBRSb5OdyXk1cC9wBXAO6vqE7OUryQvpLtd61DgCLp1Wi4Gjq+qIR5/LEmSJG20hr4ycj/gPr/c04WGmdurfgj8x9o0XlXfAA5fi3qfp7uqspA6y4BlC6yzlO7xxfMtfzdwUv+SJEmSNilDP9p30VzHkjyS7tanbegWPpQkSZK0CVuvE9hHVdV36OZIPBR464bqV5IkSdLGaYOFEfjlI20/C7xwQ/YrSZIkaeOzQcNI727gwQ36lSRJkrQR2aBhpF8Z/U+Af9+Q/UqSJEna+Az9aN+3rKafhwEH0T1V6y+H7FeSJEnS5Bn60b7HruH4rXRraPz3gfuVJEmSNGGGDiPPmGP/vcDNwLf6tTUkSZIkbeKGXmfkc0O2J0mSJGl6tXialiRJkiQNPoH9N9e2blV9b8ixSJIkSdq4DT1nZAVQa1GvGH4skiRJkjZiQweADwCLgH2BlcDXgB/RLXL4eLrH+n6OLrRIkiRJ2oQNHUb+FvgCcBJwXFXdOnMgyQOA44CXAv+lqq4ZuG9JkiRJE2ToCezvAL5eVUeNBhGAqrq1qo4E/q0vJ0mSJGkTNnQY2Re4dA1lLgWePnC/kiRJkibM0GFkS7r5IauzS19OkiRJ0iZs6DDyVeDgJE+Y7WCSfYAXAFcM3K8kSZKkCTP0BPbjgPOBLyb5R+Bi4MfAb9DdmvWndAHouIH7lSRJkjRhBg0jVfUvSQ4G/h54GXDIyOEANwOvrqp/HbJfSZIkSZNn8IUGq+qcJJ8CDgL2pltbZCXdrVkfr6rbh+5TkiRJ0uRZL6ue94Hjn/qXJEmSJN3H0BPY/5MkOyZ52PrsQ5IkSdJkGjyMJNk2ybuS/Ai4Ebh+5NjvJDkvyd5D9ytJkiRpsgwaRpJsD3wBOBL4IXAV3cT1GV8Hnga8cMh+JUmSJE2eoa+MvBl4DPCyqtob+Mjowaq6A/gcsP/A/UqSJEmaMEOHkecCn66qD6ymzHeBhw7cryRJkqQJM3QY2RW4cg1lbqN73K8kSZKkTdjQYeTnwIPWUObhdBPbJUmSJG3Chg4jXwEOTLLdbAeT7AL8IXDpwP1KkiRJmjBDh5H3ADsD5yXZc/RA//4jwFbAyQP3K0mSJGnCDLoCe1V9OslxwFuBbwD/AZDkRmBHusf8/nlVLR+yX0mSJEmTZ/BFD6vqOLpH9/5v4GbgHqCA84D/p6reOXSfkiRJkibPoFdGkuwL3FpVFwIXDtm2JEmSpOky9JWRC4FXD9ymJEmSpCk0dBi5EVg1cJuSJEmSptDQYeQiYPHAbUqSJEmaQkOHkWOARyf56yS/NnDbkiRJkqbIoBPYgb+ke6Tvm4BXJPk/wI/onqY1qqrqFQP3LUmSJGmCDB1GXjby3w/uX7MpwDAiSZIkbcKGDiMPH7g9SZIkSVNqncNIkpcCX6uqK6vquwOMSZIkSdImYIgJ7EuB54zuSHJIkgsGaFuSJEnSlBr6aVozFgFPX09tS5IkSZoC6yuMSJIkSdJqGUYkSZIkNWEYkSRJktTEUGFkfFFDSZIkSVqtodYZOTbJseM7k9wzR/mqqqHXOJEkSZI0QYYKBFnP5SVJkiRNmXUOI1XlvBNJkiRJC2aQkCRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDUxUWEkyc5JXpnk3CTfSbIqycoklyZ5RZJZP0+SxUnOS3JTX+fKJK9Pstlq+jowyUV9+7cl+VKSQ9YwvkOSfLkvv7Kvf+Bqym+W5Mh+PKv68Z2XZPH8z4okSZI0mSYqjADPB04Hfgf4EvBu4KPAY4H/CXw4SUYrJDkIuBjYFzgXOBXYAjgJOHu2TpIcDizr2z2r7/MhwNIkJ8xR5wRgKbBLX/4sYC9gWd/eePn0/Z/Yj+fUfnz7Ahf345YkSZKm1uatB7BA1wDPBj5ZVffO7EzyJuDLwPOA59IFFJI8gC4Y3APsV1WX9fv/CrgAWJLk4Ko6e6StRcAJwE3AE6tqRb//bcBXgKOSfLSqvjBSZzFwFHAt8KSqurnf/07gcuCEJJ+Yaat3MLAEWA7sX1V39nVOAy4FTk9yQVX9fB3PmSRJkrRRmqgrI1V1QVUtGw0i/f4fAaf1b/cbObQEeCBw9kwQ6cvfCRzTv33tWDeHAlsCp46Ghz5g/E3/9jVjdWbev30miPR1VgDv7dt7+VidmX6PmQkifZ2vAB/qx70ESZIkaUpNVBhZg//ot3eP7Htmvz1/lvIXA3cAi5NsOc86nxors1Z1kmwFLO77v2QB/UiSJElTY9Ju05pVks2Bl/ZvRwPBo/vtNeN1quruJNcDjwEeAVw1jzo3JLkd2DXJ1lV1R5JtgIcCt1XVDbMM79v9dveRfbsBmwHXVdXd960ya505Jbl8jkN7zKe+JEmS1MK0XBl5B91k8/Oq6tMj+7fvtyvnqDezf4e1qLP92HZ99LHDHMclSZKkiTfxV0aSvI5u8vi3gJc0Hk4TVbXPbPv7KyZ7b+DhSJIkSfMy0VdG+kfmvgf4JvCMqrpprMj4VYxxM/tvWYs6K8e266OPW+Y4LkmSJE28iQ0jSV4PnAJ8gy6I/GiWYlf32/vMvejnmTycbsL7dfOsswuwDfD9qroDoKpuB34AbNsfH/eofjs6B+VauscNP6Ifx3zqSJIkSVNlIsNIkj+nW7Twa3RB5CdzFL2g3x4wy7F9ga2B5VV11zzr/MFYmbWq0z/Kd3nf/9MW0I8kSZI0NSYujPQLFr6DbjHB/avqxtUUPwe4ETg4yRNH2tgKOL5/+76xOmcCdwGH9wsgztTZEXhT//a0sToz79/cl5upswg4rG/vzLE6M/0e349nps6TgBcAP6VfvFGSJEmaRhM1gT3JIcDb6G5xugR4XZLxYiuqailAVd2a5FV0oeSiJGfTraz+bLpH+J5Dt8DgL1XV9UneCJwMXJbkQ8Av6BYg3BV41+jq632d5UlOBN4AXJnkHGALulCxE3DE2OrrAGfTrRa/BPhqkmXAzn2dzYBXVdWtCz5JkiRJ0oSYqDBCN8cDul/WXz9Hmc8BS2feVNXHkjwdeDPwPGAr4Dt0weHkqqrxBqrqlCQrgKPp1i+5H90k+WOq6v2zdVpVRyX5Ot2VkFcD9wJXAO+sqk/MUr6SvJDudq1DgSOAO575EhgAABTtSURBVOkWYzy+qpbPfRokSZKkyTdRYaSqjgWOXYt6nwf+cIF1lgHLFlhnKSNBaB7l76ab+3LSQvqRJEmSpsHEzRmRJEmSNB0MI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpiYkLI0mWJDklySVJbk1SSc5aQ53FSc5LclOSVUmuTPL6JJutps6BSS5KsjLJbUm+lOSQNfRzSJIv9+VX9vUPXE35zZIc2Y9nVT++85IsXvOZkCRJkibbxIUR4BjgcODxwA/WVDjJQcDFwL7AucCpwBbAScDZc9Q5HFgGPBY4CzgdeAiwNMkJc9Q5AVgK7NKXPwvYC1jWtzdePn3/J/bjObUf377Axf24JUmSpKk1iWHkSGB34AHAa1dXMMkD6ILBPcB+VfWKqnojXZD5ArAkycFjdRYBJwA3AU+sqsOq6kjgccC1wFFJnjJWZzFwVH/8cVV1ZFUdBuzTt3NC3+6og4ElwHLg8VX1xqp6BfCMfrynJ9luvidFkiRJmjQTF0aq6sKq+nZV1TyKLwEeCJxdVZeNtHEn3RUWuG+gORTYEji1qlaM1LkZ+Jv+7WvG6sy8f3tfbqbOCuC9fXsvH6sz0+8x/Xhm6nwF+FA/7iVr/ISSJEnShJq4MLJAz+y3589y7GLgDmBxki3nWedTY2XWqk6SrYDFff+XLKCfWSW5fLYXsMd86kuSJEktTHsYeXS/vWb8QFXdDVwPbA48Yp51bgBuB3ZNsjVAkm2AhwK39cfHfbvf7j6ybzdgM+C6fhzzqSNJkiRNlc1bD2A9277frpzj+Mz+HRZYZ5u+3B3rsY/xOnOqqn1m299fHdl7Pm1IkiRJG9q0XxmRJEmStJGa9jAyc4Vh+zmOz+y/ZS3qrBzbro8+bpnjuCRJkjTxpj2MXN1v7zP3IsnmwMOBu4Hr5llnF7pbtL5fVXcAVNXtdOudbNsfH/eofjs6B+Vausf3PqIfx3zqSJIkSVNl2sPIBf32gFmO7QtsDSyvqrvmWecPxsqsVZ3+Ub7L+/6ftoB+JEmSpKkx7WHkHOBG4OAkT5zZ2T9a9/j+7fvG6pwJ3AUcPrpQYZIdgTf1b08bqzPz/s19uZk6i4DD+vbOHKsz0+/x/Xhm6jwJeAHwU+Cja/h8kiRJ0sSauKdpJXkO8Jz+7YP77VOSLO3/+8aqOhqgqm5N8iq6UHJRkrPpVkR/Nt0jfM+hW2Dwl6rq+iRvBE4GLkvyIeAXdAsQ7gq8q6q+MFZneZITgTcAVyY5B9iCLlTsBBwxuoBi72zguX27X02yDNi5r7MZ8KqqunUtTpEkSZI0ESYujACPBw4Z2/cIfrVWyHeBo2cOVNXHkjwdeDPwPGAr4Dt0weHk2VZyr6pTkqzo23kp3RWkb9Ktlv7+2QZVVUcl+TrdlZBXA/cCVwDvrKpPzFK+kryQ7natQ4EjgDvpFmM8vqqWr/lUSJIkSZNr4sJIVR0LHLvAOp8H/nCBdZYByxZYZymwdAHl7wZO6l+SJEnSJmXa54xIkiRJ2kgZRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYaSxJLsmOSPJD5PclWRFkncn2bH12CRJkqT1afPWA9iUJdkNWA48CPg48C3gt4H/ChyQ5KlV9bOGQ5QkSZLWG6+MtPV3dEHkdVX1nKr6i6p6JnAS8Gjg7U1HJ0mSJK1HhpFG+qsizwJWAO8dO/xW4HbgJUm22cBDkyRJkjYIw0g7z+i3n6mqe0cPVNXPgc8DWwNP3tADkyRJkjYE54y08+h+e80cx79Nd+Vkd+BfV9dQksvnOPRbV111Ffvss8/ajXAd3PCDlRu8z9b2+exbWg9BkiRpg7vqqqsAFq1NXcNIO9v327l+a5/Zv8M69HHPqlWrVl5xxRUr1qGNtbFHv/3WBu63qSt+3HoEwCZ67jcCnvc2PO/teO7b8Ly347lfvUXArWtT0TAyBapqw1/6WI2ZKzUb27g2BZ77NjzvbXje2/Hct+F5b8dzv/44Z6SdmSsf289xfGb/LRtgLJIkSdIGZxhp5+p+u/scxx/Vb+eaUyJJkiRNNMNIOxf222cl+U//Dkm2A54K3AF8cUMPTJIkSdoQDCONVNW1wGfoJvwcNnb4OGAb4INVdfsGHpokSZK0QTiBva0/A5YDJyfZH7gK+B26NUiuAd7ccGySJEnSepWqaj2GTVqShwFvAw4AdgZuAM4Fjquqm1uOTZIkSVqfDCOSJEmSmnDOiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxojZLsmuSMJD9McleSFUnenWTHBbazU19vRd/OD/t2d11fY59063ruk2yT5EVJ/inJt5LcnuTnSS5LclSSLdb3Z5hEQ33Nj7W5b5J7klSS44cc7zQZ8twn2bv/2v9+39aPk3wuyUvXx9gn2YDf5383ycf7+ncm+V6S85IcsL7GPqmSLElySpJLktzaf284ay3bGvx71jQb4twn2TnJK5Ocm+Q7SVYlWZnk0iSvSOLv2PPkOiNarSS70a0S/yDg48C3gN+mWyX+auCpVfWzebSzc9/O7sAFwFeAPYCDgJ8AT6mq69bHZ5hUQ5z7/heATwE3ARcC3wF2BJ4NPLhvf/+qunM9fYyJM9TX/Fib2wFXAr8ObAu8vaqOGXLc02DIc5/kcOA9wM3AJ4EfADsBjwW+X1UHD/4BJtSA3+dfC/wdcDvd4r3fB3YFngtsDRxTVW9fH59hEiX5GvBbwG1052oP4B+r6sULbGfw71nTbohzn+Q1wPvoFqu+EPge8Bt0X+/bAx8Fnl/+or1mVeXL15wv4NNAAUeM7T+x33/aPNv5+778u8b2v67ff37rz7qxvYY498DjgRcBW4zt3w64vG/nqNafdWN6DfU1P1b3DLpA+Ka+jeNbf86N8TXg95tnAff27W03y/Ffa/1ZN6bXQN9rfg24BVgFPHrs2J7AncAdwJatP+/G8qILC48CAuzXn+uzWvz7bWqvIc498Ezgj4H7je1/MF0wKeB5rT/rJLy8MqI59X9t+Q6wAtitqu4dObYd3V8DAjyoqm5fTTvb0l39uBfYpap+PnLsfsB1wP/V9+HVEYY792vo40+BfwQ+UVV/vM6DngLr47wnOQj4GPASYHPgTLwych9Dnvsk/wd4JPCb5V+EV2vA7/O/AfwIuLKqfmuW41cCewG/7r/JfSXZj+6v6wv96/x6/1kx7db23K+hzTcBbwdOraojhmhzmnk/m1bnGf32M6Pf4AD6QPF5ukvvT15DO08G7g98fjSI9O3M/PVytD8Nd+5X5z/67d3r0Ma0GfS8J3kQcDrwsapaq3vBNyGDnPskjwUeB3wGuCnJM5Ic3c+R2t/7uO9jqK/5nwA/BXZP8qjRA0l2p/sr9NcMIoPbED8rtHD+fF0AvylrdR7db6+Z4/i3++3uG6idTcmGOGeH9tvz16GNaTP0eT+d7vvsa9ZlUJuIoc79k/rtT4CL6OaovRM4AfgX4GtJHrn2w5w6g5z36m6zOIzu6/3yJO9P8rdJPkB3S+i/Ac8fYLz6z/z5upFJsjkw85AMf77Ow+atB6CN2vb9duUcx2f277CB2tmUrNdz1k/uPQD4Gt18BnUGO+9JDqV7UMALqurHA4xt2g117h/Ub19BN2n9j4BL6SaWvgV4MfDJJHtV1S/WfrhTY7Cv+ar6SJIfAv/Mr34ZA/gx3e2J3oY7PH++bnzeQfegjPOq6tNrKiyvjEibnCTPBd5Nd3/386rqP9ZQRQuUZBHdOf5IVX247Wg2OTM/1zYDDq6q86rq1qr6Nt0vyJfR/ZX4ea0GOK2SvJju6tMldJPWt+63/wqcCpzdbnTS+pfkdcBRdE80e0nj4UwMw4hWZ+YvKtvPcXxm/y0bqJ1NyXo5Z0meQ/cLwU+A/XxgwH0Mdd7PoHuq0J8NMahNxFDnfub4j6rqC6MH+luJPt6//e0Fj3A6DXLe+3khZ9DdjvWSqvpWVa2qqplfyi4Hnt9PFtZw/Pm6kRh5nPg3gWdU1U2NhzQxDCNanav77Vz3ms5MUpzrXtWh29mUDH7Okjwf+AjdLRNPr6qr11BlUzTUed+b7nahn/aLaVWSortVBeDN/b6Prdtwp8rQ32/m+uXr5n57/3mOa9oNdd6fRfd438/NMpH6XuDi/u0+azNIzcmfrxuBJK8HTgG+QRdEftR4SBPFOSNanQv77bOS3G+WRwY+le658V9cQztfpPsr8VOTbDfLo32fNdafhjv3M3VeBLyf7h76Z3hFZE5DnfcP0N2iMu5RwL50c3UuB766ziOeHkN+v7kdWJRkm1keZ/rYfnv9AGOeBkOd9y377QPnOD6z33k6wxr0Z4UWLsmf080T+Rrwe1V1Y+MhTRyvjGhOVXUt3eMxF9E9JWXUccA2wAdHf9gn2SPJHmPt3AZ8sC9/7Fg7h/ftf9pfkH9lqHPf7z+E7pfj7wH7ep7nNuDX/Ouq6pXjL351ZeST/b73rrcPM2EGPPd3AP8AbAUcnyQj5fcCXkb3uM1zhv8Uk2fA7zWX9NslSR43eiDJ44EldIvAXTDc6DcdSX6tP++7je5fm38/Lcxc574/9ld0QeRyYH+DyNpx0UOtVv8/33K6W04+DlwF/A7ds82vARaPPje+vxWFqspYOzv37exO98Poy3QTGw+im7+wuP+mqt4Q5z7JM+gmlN6P7n7uf5+lq1uq6t3r6WNMnKG+5udo+2W46OGcBvx+8wDgc8DjgS/RrbXwG8Bz6W7Pen1VvWd9f55JMeB5PwN4Od3Vj3OB79L9kvwcYAvg3VV15Hr+OBOjn8P3nP7tg4Hfp3vi2Eywu7Gqju7LLqK7mvfdqlo01s6C/v00zLnv/9C3FLiH7hat2Z5otqKqlg49/qmzpiXaffkCHkb3C9QNdD9kvkv3pKAdZylb9PNEZzm2E93kru/27dxA9wvyrq0/48b6WtdzT/dX4FrDa0Xrz7mxvYb6mp+l7My/x/GtP+PG+hrw+822dCsgXwPcRTeH5DPAs1p/xo3xNcR5p1vp+2V067vcTHcF6ia6p2kd3PozbmwvujsF5vW9mS7Uzfn9eiH/fr6GOffzaKOAi1p/1kl4eWVEkiRJUhPOGZEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRpAiVZkuSUJJckuTVJJTlroLb369tb0+th69LP5kMMVpIkSdIGdwzwW8BtwPeBPQZsewVw3BzH9gKeC3yjqv59XToxjEiSJEmT6Ui6EPId4OnAhUM1XFUrgGNnO5bkn/v/PH1d+/E2LUmSJGkCVdWFVfXtqqr51knywiQXJrklyZ1JrkpyTJIt51n/14E/AVYBH1jLof+SYUSSJEnaBCQ5A/gn4JHAR4H3AjcBfw2cn2Q+d00dAmwJfKSqblnXMXmbliRJkjTlkrwMeDlwLvCiqlo1cuxY4K3AYcB71tDUq/rt3w8xLq+MSJIkSdPvvwJ3A4eOBpHeXwM/A160ugaSPB14NN3E9eVDDMorI5IkSdIUS7I13VO3bgRen2S2YncBe66hqVf32/8x1NgMI5IkSdJ02xEI8EC627EWLMlOwPPoJq5/cKiBeZuWJEmSNN1W9tuvVlVW91pNGzMT1z88xMT1GYYRSZIkaYpV1W3AvwGP6a9wrI2ZieuD3aIFhhFJkiRpU3AisAVwRpIdxg8m2THJ3rNVTPI0uvkkg01cn+GcEUmSJGkCJXkO8Jz+7YP77VOSLO3/+8aqOhqgqs5Isg/wZ8C1ST4NfA/YCXg4sC9wJvCaWboafOL6Lz/DAhZslCRJkrSRGFkfZC7frapFY3UOpAscvw3sQLfo4feAzwBnVdW3xsrvCPwQKOAhQ84XAcOIJEmSpEacMyJJkiSpCcOIJEmSpCYMI5Ik6f9vv44FAAAAAAb5Ww9jT1kEsJARAABgISMAAMBCRgAAgIWMAAAACxkBAAAWMgIAACxkBAAAWMgIAACwkBEAAGAhIwAAwEJGAACAhYwAAAALGQEAABYyAgAALAJDFH9j7zQ+ggAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 260,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAIJCAYAAAC2k+IhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3dfZxlVX3n+8/XJoCA8jQaUMy0EhGukihgYtqRB7kxaIgYba+YRBGMjg5gROAmAolgMHFGBAWM5JJIG0nSJDjotCJqAshD+wToRUcEeWiNilEEGqEbTMNv/ti7zMmhqrqq+1SvOtWf9+t1Xpuz9/qttap2H1797X32XqkqJEmSJGlTe0zrCUiSJEnaPBlGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVITW7SegOZOkjuAxwOrGk9FkiRJC9di4L6qeupsCw0jC9vjH/vYx+6011577dR6IpIkSVqYbrrpJtauXbtBtYaRhW3VXnvttdP111/feh6SJElaoPbdd19uuOGGVRtS6z0jkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWpii9YT0MK0+I8+2XoKm9yqd/9m6ylIkiSNFa+MSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmhi7MJJkVZKa4vWDKWqWJLk0yd1J1ia5MclbkyyaZpxDk1yZZHWS+5N8MckR65nbEUm+1Ldf3dcfOk37RUmO6+eztp/fpUmWzPw3IkmSJI2nLVpPYAOtBt43yf77h3ckOQz4KPAgcBFwN/BbwFnA84FXTlJzDHAO8GPgQuCnwFJgWZK9q+qESWrOAI4HvgucD2wJHA6sSHJsVZ071D7A8r7fm4FzgZ2AVwFXJXlFVX18vb8JSZIkaUyNaxi5t6pOXV+jJI+nCwYPAwdW1XX9/j8GLgeWJjm8qpYP1CwGzqALLftV1ap+/zuBLwPHJ/loVX1+oGYJXRC5DXhuVd3T738PcD1wRpJPTPTVO5wuiKwEDq6qB/ua84BrgPOTXF5VP5ndr0aSJEkaD2P3Na1ZWgo8AVg+EUQA+r/4n9K/ffNQzVHAVsC5g+GhDxh/1r9901DNxPt3TQSRvmYV8IG+vyOHaibGPWUiiPQ1X6a7gvOEfv6SJEnSgjSuYWSrJL+X5KQkf5DkoCnu/3hhv71skmNXAWuAJUm2mmHNp4babFBNkq2BJf34V89iHEmSJGnBGNevae0CfGRo3x1Jjqyqzw3se0a/vWW4g6pal+QO4JnA04CbZlBzZ5IHgN2SbFNVa5JsCzwZuL+q7pxkrt/qt3sM7NsdWATcXlXrZlgzpSTXT3Foz5nUS5IkSS2M45WRC4CD6QLJtsDewF8Ci4FPJfnlgbbb99vVU/Q1sX+HDajZfmg7F2PsMMVxSZIkaeyN3ZWRqjptaNfXgTcluZ/uJvJTgd/e1PNqqar2nWx/f8Vkn008HUmSJGlGxvHKyFTO67f7D+wbvooxbGL/vRtQs3poOxdj3DvFcUmSJGnsLaQw8qN+u+3Avpv77aPuvUiyBfBUYB1w+wxrdu37/25VrQGoqgeA7wHb9ceHPb3fDt6Dchvd44af1s9jJjWSJEnSgrKQwsjz+u1gsLi83x4ySfv9gW2AlVX10AxrXjzUZoNq+kf5ruzHf8EsxpEkSZIWjLEKI0n26p9eNbx/Md0K5tCtmD7hYuAu4PAk+w203xo4vX/7waHuLgAeAo7p+52o2RE4qX973lDNxPuT+3aD8zq67++CoZqJcU/v5zNR81y6Vdh/RLdyvCRJkrQgjdsN7K+iWwH9KuDbwE/oHpP7m8DWwKV0q6cDUFX3JXkDXSi5MslyupXVX0r3CN+L6RYYZKDmjiQnAmcD1yW5CPgp3QKEuwHvHVx9va9ZmeRM4G3AjUkuBrbs57sTcOzQ6usAy4GX9/1+JckKYOe+ZhHwhqq6b0N/UZIkSdJ8N25h5Aq6EPEc4Pl092/cC1xDt+7IR6qqBguq6mNJDgBOBl5BF1pupQsOZw+372vOSbIKOAF4Ld0VpG/QrZb+4ckmVlXHJ/ka3ZWQNwKPADcA76mqT0zSvpK8mu7rWkcBxwIP0i3GeHpVrZzF70WSJEkaO2MVRvoFDT+33oaPrrsWeMksa1YAK2ZZswxYNov264Cz+pckSZK0WRmre0YkSZIkLRyGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUxNiHkSS/l6T61+9P0ebQJFcmWZ3k/iRfTHLEevo9IsmX+var+/pDp2m/KMlxSW5MsjbJ3UkuTbJkmprHJjktyc1JHkzywyT/kGSvmf8GJEmSpPE01mEkyVOAc4H7p2lzDLACeBZwIXA+8CRgWZIzpqg5A1gG7Nq3vxDYG1jR9zfcPsBy4Exgy35OlwD7A1clOWySmq2AzwJ/AtwHvB/4J+C3geuS/Op6fwGSJEnSGNui9QQ2VB8ALgB+DPxP4IRJ2iwGzgDuBvarqlX9/ncCXwaOT/LRqvr8QM0S4HjgNuC5VXVPv/89wPXAGUk+MdFX73BgKbASOLiqHuxrzgOuAc5PcnlV/WSg5m3A84GLgVdV1SN9zUXAx4APJdl7Yr8kSZK00IzzlZG3AC8EjgQemKLNUcBWwLmD4aEPGH/Wv33TUM3E+3dNBJG+ZhXwgb6/I4dq3txvT5kIIn3Nl4GLgCfQhRXgZ0FqYpz/dzBwVNXHgauB/ws4YIqfS5IkSRp7YxlG+nsq3g28v6qumqbpC/vtZZMc+9RQmw2qSbI1sARYQxciZjLO7sAvALdU1R2zmJskSZK0YIzd17SSbAF8BPgOcNJ6mj+j394yfKCq7kzyALBbkm2qak2SbYEnA/dX1Z2T9PetfrvHwL7dgUXA7VW1boY1U85rmpopJbl+ikN7zqRekiRJamHswgjdDd/PAf5LVa1dT9vt++3qKY6vBrbt262ZYXuAHWY5xihqJEmSpAVlrMJI/4Spk4D3Dt50vrmrqn0n299fMdlnE09HkiRJmpGxuWek/3rW39B9temPZ1g2cYVh+ymOD1+hmGn7ezdgjI2tkSRJkhaUsQkjwHZ091DsBTw4sNBhAe/o25zf73tf//7mfvuoey+S7Er3Fa3vVtUagKp6APgesF1/fNjT++3gvR63AQ8DT+sD00xqppzXNDWSJEnSgjJOX9N6CPjrKY7tQ3cfyTV0f9Gf+ArX5XRreRwysG/CiwfaDLoceE1fc8H6aqrqwSQrgRf0rytmMM5tdDfg75HkqZM8UWuquUmSJEkLxthcGamqtVX1+5O9gP/VN/twv++i/v0FdCHmmH4BRACS7Mi/P4nrvKGhJt6f3LebqFkMHN33NxxSPthvT+8f9TtR81zgVcCPgI8O/Cw1MM7/SPKYgZrD6ELNN4DPTf0bkSRJksbbOF0ZmbWquiPJicDZwHX96uY/pVuAcDcmuRG+qlYmOZNuhfQbk1wMbEkXKnYCjh1afR1gOfDyvt+vJFkB7NzXLALeUFX3DdWcCRza13wxyT/TrT3ySronex3l6uuSJElayBZ0GAGoqnOSrAJOAF5LdzXoG3SrpX94iprjk3yN7krIG4FHgBuA91TVJyZpX0leDaykW/X9WOBB4Crg9KpaOUnNQ0l+Hfgj4NXAccB9wMeAd1TVNzbqB5ckSZLmuQURRqrqVODUaY6vAFbMss9lwLJZtF8HnNW/Zlqzhm7dlD+ZzdwkSZKkhWBs7hmRJEmStLAYRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNTHSMJLk50bZnyRJkqSFa9RXRr6X5L8n+cUR9ytJkiRpgRl1GHkMcCJwc5LPJnlFkkUjHkOSJEnSAjDqMPIk4PeAq4GDgX8AvpvkXUkWj3gsSZIkSWNspGGkqn5aVX9XVQcCewLvA7YA3g7cmuTSJIcl8cZ5SZIkaTM3Z6Ggqm6pquOBJ/PvV0sOAf4n8J0kpyZ50lyNL0mSJGl+m/MrFFX1U+CTwCXA94HQfZ3rT4A7krwvyVZzPQ9JkiRJ88uchpEkz0tyAV0IOQvYFjgbeDZwFHAzcCzd17kkSZIkbUa2GHWHSR4HvAb4r8Cz6K6EfAX4C+Dvqmpt3/TGJB8BLgOWAm8e9VwkSZIkzV8jDSNJ/hr4f4BtgIeAjwB/UVVfmqx9VT2c5ErghaOchyRJkqT5b9RXRo4EbgPOAy6oqrtnUHMl8M4Rz0OSJEnSPDfqMHJIVX1mNgVVdS1w7YjnIUmSJGmeG/U6I7MKIpIkSZI2XyMNI0kOTvKhqdYPSfKk/viBoxxXkiRJ0vgZ9de0jgX2rKrvT3awqr6f5NeA7enuFZEkSZK0mRr1OiP7ACvX0+YaYL8RjytJkiRpzIw6jDyRboHD6fxr306SJEnSZmzUYWQ18JT1tHkK8MCIx5UkSZI0ZkYdRr4EvCzJLpMd7G9sf1nfTpIkSdJmbNRh5BzgccDVSV6aZCuAJFslOQy4CtgOOHvE40qSJEkaMyN9mlZVfSbJnwJ/DFwCVJJ7gB2B9K8/rarLRjmuJEmSpPEz6isjVNU7gEOAS4G76R7jezfwSeA3+uOSJEmSNnOjXmcE+NlK7K7GLkmSJGlKI78yIkmSJEkzMSdXRgCSbEN3r8iiyY5X1XfmamxJkiRJ89/Iw0iS1wB/COw1TbOai7ElSZIkjY+RBoIkrwM+BDwMXA38C7BulGNIkiRJWhhGfXXiBOAe4L9U1U0j7luSJEnSAjLqG9h/EfhHg4gkSZKk9Rl1GLkbeGjEfUqSJElagEYdRj4BHJgkI+5XkiRJ0gIz6jDydmAr4Lwk2424b0mSJEkLyKhvYP9HYA3w+8DvJPkWcO8k7aqqDh7x2JIkSZLGyKjDyIED/70t8Owp2tWIx5UkSZI0ZkYaRqpq1F/7kiRJkrRAGR4kSZIkNWEYkSRJktTEyMNIksckOTbJF5KsTrJu4NhzkvxFkj02ov//nuSfk/xLkrVJ7k7ylSTvSLLzFDVLklzat12b5MYkb02yaJpxDk1yZf8z3J/ki0mOWM/cjkjypb796r7+0GnaL0pyXD+fiZ/l0iRLZv4bkSRJksbTSMNIki2BzwLvA3YHfgIMrjlyB3AU8LsbMcxxdDfHfxZ4P/C3wDrgVODGJE8ZmtNhwFXA/sAlwLnAlsBZwPIpfo5jgBXAs4ALgfOBJwHLkpwxRc0ZwDJg1779hcDewIq+v+H26cc/s5/Puf389geu6uctSZIkLVijvjJyInAQcBrw88BfDR6sqnvpgsFvbMQYj6+q51XVUVX1R1V1bFU9F/gzusDw9omGSR5PFwweBg6sqtdX1Yl0T/n6PLA0yeGDnSdZDJxBt5r8flV1dFUdB/wScBtwfJJfG6pZAhzfH/+lqjquqo4G9u37OaPvd9DhwFJgJfDsqjqxql5P9/t7GDg/yeM24vckSZIkzWujDiO/C1xbVe+sqkeY/BG+dwC/sKEDVNWDUxz6h3779IF9S4EnAMur6rqhPk7p3755qJ+j6BZuPLeqVg3U3EMXeADeNFQz8f5dfbuJmlXAB/r+jhyqmRj3lMGfqaq+DFzUz3vpJD+nJEmStCCMOow8FfjCetrcDew04nEBfqvf3jiw74X99rJJ2l9Ft0DjkiRbzbDmU0NtNqgmydbAkn78q2cxjiRJkrRgjHrRwweBHdbT5heYfFX2WUlyArAdsD2wH/Bf6ILIuweaPaPf3jJcX1XrktwBPBN4GnDTDGruTPIAsFuSbapqTZJtgScD91fVnZNM9Vv9dvCm/d2BRcDtVbXu0SWT1kwpyfVTHNpzJvWSJElSC6MOI18FXpRky6r66fDBJNvT3S+ycgRjnUB3X8qEy4DXVdWPBvZt329XT9HHxP7BADWTmm37dmvmcIzhGkmSJGlBGfXXtP4/4CnA3/Y3j/9Mkh3onja1I3Dexg5UVbtUVYBdgJfTXd34SpJ9NrbvcVNV+072Ar7Zem6SJEnSVEZ6ZaSq/j7JrwOvA14K3AOQ5Dq6r0NtBXygqi4d4Zj/ClyS5Aa6r1b9Dd0jeeHfrzBsP1ntwP7Br42tBv5Tf+zH09SsHtrOdozZ1kiSJEkLysgXPayqo+ieSPUNuidCBdgHuBV4fVUdO+ox+3G/3Y/5zCT/qd99c7991L0XSbagu+F+HXD7wKHpanal+4rWd6tqTT/uA8D3gO3648Mmnu41eA/KbXSP731aP4+Z1EiSJEkLysjDCEBVLauq59DdYL4b8Liq2ruqLpiL8QY8qd8+3G8v77eHTNJ2f2AbYGVVPTSwf7qaFw+12aCa/lG+K/vxXzCLcSRJkqQFY07CyISqWltV3++vHmy0JHv0N8EP739MkncBT6QLFxNrfVwM3AUcnmS/gfZbA6f3bz841N0FwEPAMYMLFSbZETipfzt8z8vE+5P7dhM1i4Gj+/6Gg9jEuKf385moeS7wKuBHwEeHf1ZJkiRpoRj107Tm2kuAP09yDd3iiT+me6LWAXQ3sP8AeMNE46q6L8kb6ELJlUmW061z8lK6R/heTLfAIAM1dyQ5ETgbuC7JRcBP6RYg3A14b1V9fqhmZZIzgbcBNya5GNiSLlTsBBw7uIBibzndjfdL6W68XwHs3NcsAt5QVfdt6C9KkiRJmu9GGkaS3L7+VgBUVe2+AUP8E/CLdGuKPIfu0bcP0N1b8RHg7Kq6e2igjyU5ADgZeAWwNd39K2/r2z9qlfiqOifJKrrHB7+W7grSN+hWS//wFD/Q8Um+Rncl5I3AI8ANwHuq6hOTtK8kr6b7utZRwLF067RcBZxeVaN4/LEkSZI0b436yshjgEf95Z4uNEx8ver7wL9tSOdV9XXgmA2ou5buqspsalYAK2ZZs4zu8cUzbb8OOKt/SZIkSZuVUT/ad/FUx5L8It1Xn7alW/hQkiRJ0mZsTm9gH1RVt9LdI/Fk4B2balxJkiRJ89MmCyPws0fafhZ49aYcV5IkSdL8s0nDSG8dsEuDcSVJkiTNI5s0jPQro/828C+bclxJkiRJ88+oH+37J9OM8xTgMLqnar19lONKkiRJGj+jfrTvqes5fh/dGhr/Y8TjSpIkSRozow4jB02x/xHgHuCb/doakiRJkjZzo15n5HOj7E+SJEnSwtXiaVqSJEmSNPIb2H9hQ2ur6jujnIskSZKk+W3U94ysAmoD6orRz0WSJEnSPDbqAPA3wGJgf2A18FXgB3SLHD6b7rG+n6MLLZIkSZI2Y6MOI38OfB44Czitqu6bOJDk8cBpwGuB/1pVt4x4bEmSJEljZNQ3sL8b+FpVHT8YRACq6r6qOg743307SZIkSZuxUYeR/YFr1tPmGuCAEY8rSZIkacyMOoxsRXd/yHR27dtJkiRJ2oyNOox8BTg8yXMmO5hkX+BVwA0jHleSJEnSmBn1DeynAZcBX0jyt8BVwL8CP0/31azfoQtAp414XEmSJEljZqRhpKr+KcnhwF8CrwOOGDgc4B7gjVX1z6McV5IkSdL4GflCg1V1cZJPAYcB+9CtLbKa7qtZH6+qB0Y9piRJkqTxMyernveB4+/6lyRJkiQ9yqhvYP8PkuyY5ClzOYYkSZKk8TTyMJJkuyTvTfID4C7gjoFjv5rk0iT7jHpcSZIkSeNlpGEkyfbA54HjgO8DN9HduD7ha8ALgFePclxJkiRJ42fUV0ZOBp4JvK6q9gH+cfBgVa0BPgccPOJxJUmSJI2ZUYeRlwOfrqq/mabNt4Enj3hcSZIkSWNm1GFkN+DG9bS5n+5xv5IkSZI2Y6MOIz8BnrieNk+lu7FdkiRJ0mZs1GHky8ChSR432cEkuwIvAa4Z8biSJEmSxsyow8j7gZ2BS5PsNXigf/+PwNbA2SMeV5IkSdKYGekK7FX16SSnAe8Avg78G0CSu4Ad6R7z+4dVtXKU40qSJEkaPyNf9LCqTqN7dO//Au4BHgYKuBT4v6vqPaMeU5IkSdL4GemVkST7A/dV1RXAFaPsW5IkSdLCMuorI1cAbxxxn5IkSZIWoFGHkbuAtSPuU5IkSdICNOowciWwZMR9SpIkSVqARh1GTgGekeRPk/zciPuWJEmStICM9AZ24O10j/Q9CXh9kv8f+AHd07QGVVW9fsRjS5IkSRojow4jrxv4713612QKMIxIkiRJm7FRh5Gnjrg/SZIkSQvURoeRJK8FvlpVN1bVt0cwJ0mSJEmbgVHcwL4MeNngjiRHJLl8BH1LkiRJWqBG/TStCYuBA+aob0mSJEkLwFyFEUmSJEmalmFEkiRJUhOGEUmSJElNjCqMDC9qKEmSJEnTGtU6I6cmOXV4Z5KHp2hfVTXqNU4kSZIkjZFRBYLMcXtJkiRJC8xGh5Gq8r4TSZIkSbNmkJAkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1MVZhJMnOSX4/ySVJbk2yNsnqJNckeX2SSX+eJEuSXJrk7r7mxiRvTbJomrEOTXJl3//9Sb6Y5Ij1zO+IJF/q26/u6w+dpv2iJMf181nbz+/SJEtm/luRJEmSxtNYhRHglcD5wK8CXwTeB3wUeBbwV8A/JMlgQZLDgKuA/YFLgHOBLYGzgOWTDZLkGGBF3++F/ZhPApYlOWOKmjOAZcCuffsLgb2BFX1/w+3Tj39mP59z+/ntD1zVz1uSJElasLZoPYFZugV4KfDJqnpkYmeSk4AvAa8AXk4XUEjyeLpg8DBwYFVd1+//Y+ByYGmSw6tq+UBfi4EzgLuB/apqVb//ncCXgeOTfLSqPj9QswQ4HrgNeG5V3dPvfw9wPXBGkk9M9NU7HFgKrAQOrqoH+5rzgGuA85NcXlU/2cjfmSRJkjQvjdWVkaq6vKpWDAaRfv8PgPP6twcOHFoKPAFYPhFE+vYPAqf0b988NMxRwFbAuYPhoQ8Yf9a/fdNQzcT7d00Ekb5mFfCBvr8jh2omxj1lIoj0NV8GLurnvRRJkiRpgRqrMLIe/9Zv1w3se2G/vWyS9lcBa4AlSbaaYc2nhtpsUE2SrYEl/fhXz2IcSZIkacEYt69pTSrJFsBr+7eDgeAZ/faW4ZqqWpfkDuCZwNOAm2ZQc2eSB4DdkmxTVWuSbAs8Gbi/qu6cZHrf6rd7DOzbHVgE3F5V6x5dMmnNlJJcP8WhPWdSL0mSJLWwUK6MvJvuZvNLq+rTA/u377erp6ib2L/DBtRsP7SdizF2mOK4JEmSNPbG/spIkrfQ3Tz+TeA1jafTRFXtO9n+/orJPpt4OpIkSdKMjPWVkf6Rue8HvgEcVFV3DzUZvooxbGL/vRtQs3poOxdj3DvFcUmSJGnsjW0YSfJW4Bzg63RB5AeTNLu53z7q3ov+PpOn0t3wfvsMa3YFtgW+W1VrAKrqAeB7wHb98WFP77eD96DcRve44af185hJjSRJkrSgjGUYSfKHdIsWfpUuiPxwiqaX99tDJjm2P7ANsLKqHpphzYuH2mxQTf8o35X9+C+YxTiSJEnSgjF2YaRfsPDddIsJHlxVd03T/GLgLuDwJPsN9LE1cHr/9oNDNRcADwHH9AsgTtTsCJzUvz1vqGbi/cl9u4maxcDRfX8XDNVMjHt6P5+JmucCrwJ+RL94oyRJkrQQjdUN7EmOAN5J9xWnq4G3JBlutqqqlgFU1X1J3kAXSq5MspxuZfWX0j3C92K6BQZ/pqruSHIicDZwXZKLgJ/SLUC4G/DewdXX+5qVSc4E3gbcmORiYEu6ULETcOzQ6usAy+lWi18KfCXJCmDnvmYR8Iaqum/WvyRJkiRpTIxVGKG7xwO6v6y/dYo2nwOWTbypqo8lOQA4GXgFsDVwK11wOLuqariDqjonySrgBLr1Sx5Dd5P8KVX14ckGrarjk3yN7krIG4FHgBuA91TVJyZpX0leTcd6UNkAABNdSURBVPd1raOAY4EH6RZjPL2qVk79a5AkSZLG31iFkao6FTh1A+quBV4yy5oVwIpZ1ixjIAjNoP06untfzprNOJIkSdJCMHb3jEiSJElaGAwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKkJw4gkSZKkJgwjkiRJkpowjEiSJElqwjAiSZIkqQnDiCRJkqQmDCOSJEmSmjCMSJIkSWrCMCJJkiSpCcOIJEmSpCYMI5IkSZKaMIxIkiRJasIwIkmSJKmJsQsjSZYmOSfJ1UnuS1JJLlxPzZIklya5O8naJDcmeWuSRdPUHJrkyiSrk9yf5ItJjljPOEck+VLffnVff+g07RclOa6fz9p+fpcmWbL+34QkSZI03sYujACnAMcAzwa+t77GSQ4DrgL2By4BzgW2BM4Clk9RcwywAngWcCFwPvAkYFmSM6aoOQNYBuzat78Q2BtY0fc33D79+Gf28zm3n9/+wFX9vCVJkqQFaxzDyHHAHsDjgTdP1zDJ4+mCwcPAgVX1+qo6kS7IfB5YmuTwoZrFwBnA3cB+VXV0VR0H/BJwG3B8kl8bqlkCHN8f/6WqOq6qjgb27fs5o+930OHAUmAl8OyqOrGqXg8c1M/3/CSPm+kvRZIkSRo3YxdGquqKqvpWVdUMmi8FngAsr6rrBvp4kO4KCzw60BwFbAWcW1WrBmruAf6sf/umoZqJ9+/q203UrAI+0Pd35FDNxLin9POZqPkycFE/76Xr/QklSZKkMTV2YWSWXthvL5vk2FXAGmBJkq1mWPOpoTYbVJNka2BJP/7VsxhnUkmun+wF7DmTekmSJKmFhR5GntFvbxk+UFXrgDuALYCnzbDmTuABYLck2wAk2RZ4MnB/f3zYt/rtHgP7dgcWAbf385hJjSRJkrSgbNF6AnNs+367eorjE/t3mGXNtn27NXM4xnDNlKpq38n291dH9plJH5IkSdKmttCvjEiSJEmapxZ6GJm4wrD9FMcn9t+7ATWrh7ZzMca9UxyXJEmSxt5CDyM399tH3XuRZAvgqcA64PYZ1uxK9xWt71bVGoCqeoBuvZPt+uPDnt5vB+9BuY3u8b1P6+cxkxpJkiRpQVnoYeTyfnvIJMf2B7YBVlbVQzOsefFQmw2q6R/lu7If/wWzGEeSJElaMBZ6GLkYuAs4PMl+Ezv7R+ue3r/94FDNBcBDwDGDCxUm2RE4qX973lDNxPuT+3YTNYuBo/v+LhiqmRj39H4+EzXPBV4F/Aj46Hp+PkmSJGlsjd3TtJK8DHhZ/3aXfvtrSZb1/31XVZ0AUFX3JXkDXSi5MslyuhXRX0r3CN+L6RYY/JmquiPJicDZwHVJLgJ+SrcA4W7Ae6vq80M1K5OcCbwNuDHJxcCWdKFiJ+DYwQUUe8uBl/f9fiXJCmDnvmYR8Iaqum8DfkWSJEnSWBi7MAI8GzhiaN/T+Pe1Qr4NnDBxoKo+luQA4GTgFcDWwK10weHsyVZyr6pzkqzq+3kt3RWkb9Ctlv7hySZVVccn+RrdlZA3Ao8ANwDvqapPTNK+krya7utaRwHHAg/SLcZ4elWtXP+vQpIkSRpfYxdGqupU4NRZ1lwLvGSWNSuAFbOsWQYsm0X7dcBZ/UuSJEnarCz0e0YkSZIkzVOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGGksyW5JPpTk+0keSrIqyfuS7Nh6bpIkSdJc2qL1BDZnSXYHVgJPBD4OfBP4FeAPgEOSPL+qftxwipIkSdKc8cpIW39BF0TeUlUvq6o/qqoXAmcBzwDe1XR2kiRJ0hwyjDTSXxV5EbAK+MDQ4XcADwCvSbLtJp6aJEmStEkYRto5qN9+pqoeGTxQVT8BrgW2AZ63qScmSZIkbQreM9LOM/rtLVMc/xbdlZM9gH+erqMk109x6Jdvuukm9t133w2b4Ua483urN/mYre372T9pPQVJkqRN7qabbgJYvCG1hpF2tu+3U/2tfWL/DhsxxsNr165dfcMNN6zaiD42xJ799pubeNymbvjX1jNoYrM815shz/PmwfO8+fBcbx425XleDNy3IYWGkQWgqjb9pY9pTFypmW/z0uh5rjcPnufNg+d58+G53jyMy3n2npF2Jq58bD/F8Yn9926CuUiSJEmbnGGknZv77R5THH96v53qnhJJkiRprBlG2rmi374oyX84D0keBzwfWAN8YVNPTJIkSdoUDCONVNVtwGfobvg5eujwacC2wEeq6oFNPDVJkiRpk/AG9rb+G7ASODvJwcBNwK/SrUFyC3Byw7lJkiRJcypV1XoOm7UkTwHeCRwC7AzcCVwCnFZV97ScmyRJkjSXDCOSJEmSmvCeEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRjReiXZLcmHknw/yUNJViV5X5IdZ9nPTn3dqr6f7/f97jZXc9fsjOJcJ7kySU3z2noufwZNL8nSJOckuTrJff05uXAD+xrJ/xs0N0Z1rvvzOtXn+QdzMXfNTJKdk/x+kkuS3JpkbZLVSa5J8voks/p7np/p+WuU53q+faZdgV3TSrI73SrxTwQ+DnwT+BXgD4BDkjy/qn48g3527vvZA7gcWA7sCRwJ/GaSX6uq2+fmp9BMjOpcDzhtiv3rNmqi2linAL8M3A98l+5zOGtz8OdFozeSc91bDbxvkv33b0Sf2nivBD5It2DyFcB3gJ8HXg78FfDiJK+sGSwq52d63hvZue7Nn890VfnyNeUL+DRQwLFD+8/s9583w37+sm//3qH9b+n3X9b6Z93cXyM811d2/2tp/zP5mvT8HAQ8HQhwYH9uL2z158XXWJzrVcCq1j+Pr0nPzQuB3wIeM7R/F7q/rBbwihn25Wd6Hr9GfK7n1WfaFdg1pf5fSW6l+0O7e1U9MnDscXTpPMATq+qBafrZDvgh8Aiwa1X9ZODYY4Dbgf/cj+HVkQZGda779lcCB1RV5mzCGokkB9L9C9vfVtXvzaJuZH9etGls6Lnua1cBVNXikU9McybJScC7gHOr6tj1tPUzPcZmc6779qtg/nymvWdE0zmo335m8H9MAH2guBbYBnjeevp5HvBY4NrBINL38wjdv8YMjqdNb1Tn+meSvCrJHyV5W5IXJ9lqdNNVYyP/86J5b6skv5fkpCR/kOSgJItaT0rT+rd+O5OvxvqZHm+zOdcT5s1n2ntGNJ1n9Ntbpjj+LeBFdPeB/PNG9kPfj9oY1bketHzo/Q+THF1VF2/A/DS/zMWfF81vuwAfGdp3R5Ijq+pzLSakqSXZAnht//ayGZT4mR5TG3CuJ8ybz7RXRjSd7fvt6imOT+zfYRP1o7kzynP0cbrvte5Gd0VsT+DP+9qLkhyyEfPU/OBnevNyAXAw3V9etgX2prsPcDHwqSS/3G5qmsK7gWcBl1bVp9fXGD/T42y25xrm2WfaKyOSRqqqzhradTNwUpLvA+fQBZPZ/OuNpIaqavjJeF8H3pTkfuB44FTgtzf1vDS5JG+hOy/fBF7TeDqaQxt6rufbZ9orI5rOxL+EbD/F8Yn9926ifjR3NsU5+iu677M+u78hUuPLz7QAzuu3+zedhX4myTHA+4FvAAdV1d0zLPUzPWY24lxPp8ln2jCi6dzcb6e6l+Pp/Xaq75iOuh/NnTk/R1X1IDDxAINtN7QfzQt+pgXwo37r53keSPJWuqvPX6f7y+lsFq/zMz1GNvJcT6fJZ9owoulc0W9fNLyyZ/8v288H1gBfWE8/XwDWAs8f/hfxvt8XDY2nTW9U53pKSZ4B7EgXSO7a0H40L8z5nxeNhYknK/lI9saS/CFwFvBVur+c/nCWXfiZHhMjONfTafKZNoxoSlV1G/AZuhuajh46fBpdcv7I4DPHk+yZ5D+s8ltV99M9sWFbuu8hDjqm7//TrjHSzqjOdZKnJtlpuP8kT6C7YQ5geVW5CvsYSPJz/XnefXD/hvx50fw21blOsleSR/0raZLFwLn92wvnfoaaSpI/pruJ+Xrg4Kqa8h97/EyPt1Gc6/n4mXbRQ02r/0O8Engi3VOSbgJ+le6Z5LcAS6rqxwPtC2B4wbskO/f97AFcDnwJ2As4jG5BxCX9/wzVyCjOdZLX0X3n9Bq6f1m5G/gF4CV03zm+Dvj1qvJ7x40keRnwsv7tLsBv0J2rq/t9d1XVCX3bxcAdwLeHF8ea7Z8XbXqjONdJTqW7ofUq4Nt0VzZ3B34T2Bq4FPjtqvrpnP4wmlSSI4BlwMN0X9uZ7GlYq6pqWd9+MX6mx9KozvW8/Ey3XgLe1/x/AU+h+1ftO4Gf9n943wfsOEnb6v5YTdrPTnQ3W3277+dO4EPAbq1/Rl+jOdd0jwdcBnwN+DHdQkx30/3l51hgy9Y/4+b+ors6WdO8Vg20XTy8b0P/vPgaz3MNHAD8Pd3Teu7tP9M/Aj5Lt7ZBWv+cm/NrBue4gCvXd54HjvuZnqevUZ3r+fiZ9sqIJEmSpCa8Z0SSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZKkMZRkaZJzklyd5L4kleTCORhnnyR/l+S7SR5K8q9JPpfktRvb9xajmKAkSZKkTe4U4JeB+4HvAnuOeoAkxwDvB+4BPgl8D9gJeBbwEuBvNqZ/w4gkSZI0no6jCyG3AgcAV4yy8yQvAs4GPgssraqfDB3/uY0dw69pSZIkSWOoqq6oqm9VVc20Jsmrk1yR5N4kDya5KckpSbaapPl7gLXA7wwHkX78f9uI6QNeGZEkSZI2C0k+BBxJdzXlo8C9wPOAPwUOTvLrVbWub/ss4JeAjwF3JzkI2Bco4KvAFVX1yMbOyTAiSZIkLXBJXkcXRC4Bfreq1g4cOxV4B3A03f0hAM/ttz8ErgT2H+rya0leXlW3bsy8/JqWJEmStPD9AbAOOGowiPT+FPgx8LsD+57Yb18PLAZ+E9ge2AO4ENgb+GSSLTdmUl4ZkSRJkhawJNvQPXXrLuCtSSZr9hCw18D7iYsWi4DDq+rz/fv7+kf67gnsB7wC+PsNnZthRJIkSVrYdgQCPIHu61gzcW+//cFAEAGgqirJx+nCyK+wEWHEr2lJkiRJC9vqfvuVqsp0r4Gam/vtvUzunn772I2ZmGFEkiRJWsCq6n7gfwPPTLLTDMu+ADwALE6y7STHn9Vv79iYuRlGJEmSpIXvTGBL4ENJdhg+mGTHJPtMvK+qNcBfA1sDp2fgRpMkewOvo7sh/uKNmVRmsUaKJEmSpHkiycuAl/VvdwF+A7gduLrfd1dVnTDQ/gPAfwPuBj4NfAfYCXgq3aN7L6iqNw20fzzwOeDZwBeBa4GfB15O9/Wst1bVxKOAN+xnMIxIkiRJ42dgfZCpfLuqFg/VHAq8ie7G8x3ogsl3gM8AF1bVN4fabwe8HXgl8J/pVmT/EnBGVX1mo38Gw4gkSZKkFrxnRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhOGEUmSJElNGEYkSZIkNWEYkSRJktSEYUSSJElSE4YRSZIkSU0YRiRJkiQ1YRiRJEmS1IRhRJIkSVIThhFJkiRJTRhGJEmSJDVhGJEkSZLUhGFEkiRJUhP/B5Zo3ICEznzsAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 260,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31315"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(51933, 33019)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.SUM_DEPTH <1000]),len(loci_df[loci_df.SUM_DEPTH >=75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3947"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 10]) - len(loci_df[loci_df.QUAL >= 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30502, 20105, 13880)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL <  100]), len(loci_df[loci_df.QUAL < 50]), len(loci_df[loci_df.QUAL < 30])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35404, 0)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.02])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DO NOT remove based off Fis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.SUM_DEPTH >= 75) & \n",
    "                  (df.SUM_DEPTH < 1000) & \n",
    "                  (df.QUAL >= 30) & \n",
    "                  (df.MAF >= 0.02)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22254, 25)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps\n",
      "\t--positions /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 22254 out of a possible 55771 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_bad_remove_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plink / chrom_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55771"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chroms = sorted(set([x.split(\"-\")[0] for x in loci_df.index]))\n",
    "len(chroms) # number of unique contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"chrom_map.txt\"), \"w\") as o:\n",
    "    for i, c in enumerate(chroms):\n",
    "        o.write(\"%s\\t%d\\n\" % (c, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def write_plink_files(vcf_gz):\n",
    "    !$vcftools --gzvcf {vcf_gz} \\\n",
    "    --out {vcf_gz} \\\n",
    "    --plink \\\n",
    "    --chrom-map {os.path.join(analysis_dir, \"chrom_map.txt\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf.gz\n",
      "\t--chrom-map /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/chrom_map.txt\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf.gz\n",
      "\t--plink\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Writing PLINK PED and MAP files ... \n",
      "\tRead 55772 chromosome mapping file entries.\n",
      "Done.\n",
      "After filtering, kept 22254 out of a possible 22254 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "write_plink_files(snps_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf\n",
      "\t--012\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 22254 out of a possible 22254 Sites\n",
      "Run Time = 0.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/sagebrush/denovo/SNPcall/test/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 15 out of 15 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 22254 out of a possible 22254 Sites\n",
      "Run Time = 0.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2_EW_36</td>\n",
       "      <td>21614</td>\n",
       "      <td>16.12890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2_HC_2</td>\n",
       "      <td>21403</td>\n",
       "      <td>14.45060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2_VM_5</td>\n",
       "      <td>20519</td>\n",
       "      <td>9.86954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2_YB_79</td>\n",
       "      <td>21514</td>\n",
       "      <td>18.01830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT4_DS_2</td>\n",
       "      <td>20524</td>\n",
       "      <td>23.54840</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        INDV  N_SITES  MEAN_DEPTH\n",
       "0  AT2_EW_36    21614    16.12890\n",
       "1   AT2_HC_2    21403    14.45060\n",
       "2   AT2_VM_5    20519     9.86954\n",
       "3  AT2_YB_79    21514    18.01830\n",
       "4   AT4_DS_2    20524    23.54840"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"good_snps.recode.vcf.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    15.000000\n",
       "mean     16.875816\n",
       "std       3.777805\n",
       "min       9.869540\n",
       "25%      14.480400\n",
       "50%      17.110200\n",
       "75%      18.205050\n",
       "max      24.618500\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm snps*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm fb*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
